File: LCX/ValidEndPoints/dsl/examples/market_data_dynamic.swift
import Foundation
import CommonCrypto


func fetch_all_pairs_data(pair, completion: @escaping (Result<[String: Any], Error>) -> Void) {
    // GET /api/ticker
    let urlString = "https://exchange-api.lcx.com/api/ticker"
    guard let url = URL(string: urlString) else {
        completion(.failure(NSError(domain: "Invalid URL", code: 0)))
        return
    }

    var request = URLRequest(url: url)
    request.addValue("application/json", forHTTPHeaderField: "Content-Type")


    URLSession.shared.dataTask(with: request) { data, response, error in
        if let error = error {
            completion(.failure(error))
            return
        }

        guard let data = data else {
            completion(.failure(NSError(domain: "No data", code: 0)))
            return
        }
        if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {{
            print(json)
            completion(.success(json))
        }}

    }.resume()
}

# Workflow: FetchAllPairsData (sequence)

func get_ethusdc_ticker(pair, completion: @escaping (Result<[String: Any], Error>) -> Void) {
    // GET /api/ticker
    let urlString = "https://exchange-api.lcx.com/api/ticker"
    guard let url = URL(string: urlString) else {
        completion(.failure(NSError(domain: "Invalid URL", code: 0)))
        return
    }

    var request = URLRequest(url: url)
    request.addValue("application/json", forHTTPHeaderField: "Content-Type")


    URLSession.shared.dataTask(with: request) { data, response, error in
        if let error = error {
            completion(.failure(error))
            return
        }

        guard let data = data else {
            completion(.failure(NSError(domain: "No data", code: 0)))
            return
        }
        if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {{
            print(json)
            completion(.success(json))
        }}

    }.resume()
}

func get_lcxusdc_ticker(pair, completion: @escaping (Result<[String: Any], Error>) -> Void) {
    // GET /api/ticker
    let urlString = "https://exchange-api.lcx.com/api/ticker"
    guard let url = URL(string: urlString) else {
        completion(.failure(NSError(domain: "Invalid URL", code: 0)))
        return
    }

    var request = URLRequest(url: url)
    request.addValue("application/json", forHTTPHeaderField: "Content-Type")


    URLSession.shared.dataTask(with: request) { data, response, error in
        if let error = error {
            completion(.failure(error))
            return
        }

        guard let data = data else {
            completion(.failure(NSError(domain: "No data", code: 0)))
            return
        }
        if let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] {{
            print(json)
            completion(.success(json))
        }}

    }.resume()
}

File: LCX/ValidEndPoints/dsl/examples/market_data_dynamic.scala
import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.http.scaladsl.Http
import akka.http.scaladsl.model._
import akka.http.scaladsl.model.headers.RawHeader
import scala.concurrent.Future
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import java.util.Base64


def fetch_all_pairs_data(pair): scala.concurrent.Future[String] = {
    // GET /api/ticker
    implicit val system = ActorSystem()
    implicit val materializer = ActorMaterializer()
    implicit val executionContext = system.dispatcher

    val url = "https://exchange-api.lcx.com/api/ticker"

    Http().singleRequest(
        HttpRequest(
            method = HttpMethods.GET,
            uri = url,
            headers = scala.collection.immutable.Seq(
                RawHeader("Content-Type", "application/json")
            )
        )
    ).map(response => {
        val body = response.entity.dataBytes
            .map(_.utf8String)
            .runFold("")(_ + _)
        println(body)
        body

    })
}

# Workflow: FetchAllPairsData (sequence)

def get_ethusdc_ticker(pair): scala.concurrent.Future[String] = {
    // GET /api/ticker
    implicit val system = ActorSystem()
    implicit val materializer = ActorMaterializer()
    implicit val executionContext = system.dispatcher

    val url = "https://exchange-api.lcx.com/api/ticker"

    Http().singleRequest(
        HttpRequest(
            method = HttpMethods.GET,
            uri = url,
            headers = scala.collection.immutable.Seq(
                RawHeader("Content-Type", "application/json")
            )
        )
    ).map(response => {
        val body = response.entity.dataBytes
            .map(_.utf8String)
            .runFold("")(_ + _)
        println(body)
        body

    })
}

def get_lcxusdc_ticker(pair): scala.concurrent.Future[String] = {
    // GET /api/ticker
    implicit val system = ActorSystem()
    implicit val materializer = ActorMaterializer()
    implicit val executionContext = system.dispatcher

    val url = "https://exchange-api.lcx.com/api/ticker"

    Http().singleRequest(
        HttpRequest(
            method = HttpMethods.GET,
            uri = url,
            headers = scala.collection.immutable.Seq(
                RawHeader("Content-Type", "application/json")
            )
        )
    ).map(response => {
        val body = response.entity.dataBytes
            .map(_.utf8String)
            .runFold("")(_ + _)
        println(body)
        body

    })
}

